lib/thinky.js

```diff
@@ -46,43 +48,47 @@ function Thinky(config) {
   self.Query = Query;
   self.models = {};
 
-  // Whether the database is available or not.
-  self._dbReady = false;
-  // Functions to execute once the database is ready.
-  // We do not use listeners as this used to crash node when too many listeners
-  // were added.
-  self._onDbReady = [];
-
   // Export errors
   self.Errors = Errors;
 
-  var r = self.r;
-  // Callback when the database has been created or is available.
-  var onDatabaseReady = function() {
-    self._dbReady = true;
-    for(var i=0; i<self._onDbReady.length; i++) {
-      self._onDbReady[i]();
-    }
-  }
-  r.dbList().contains(config.db).do(function(result) {
-    return r.branch(
-      result,
-      {created: 0},
-      r.dbCreate(config.db)
-    )
-  }).run().then(onDatabaseReady).error(function(error) {
-    // The `do` is not atomic, we a concurrent query could create the database
-    // between the time `dbList` is ran and `dbCreate` is.
-    if (error.message.match(/^Database `.*` already exists in/)) {
-      onDatabaseReady();
-    }
-    else {
-      // In case something went wrong here, we do not recover and throw.
-      throw error;
-    }
+  // Initialize the database.
+  self.dbReady().then().error(function(error) {
+    throw error;
   });
 }
 
+
+/**
+ * Initialize our database.
+ * @return {Promise=} Returns a promise which will resolve when the database is ready.
+ */
+Thinky.prototype.dbReady = function() {
+  var self = this;
+  if (this._dbReadyPromise) return this._dbReadyPromise;
+  this._dbReadyPromise = new Promise(function(resolve, reject) {
+    var r = self.r;
+    r.dbList().contains(self._config.db).do(function(result) {
+      return r.branch(
+        result,
+        {created: 0},
+        r.dbCreate(self._config.db)
+      );
+    }).run().then(resolve).error(function(error) {
+      // The `do` is not atomic, we a concurrent query could create the database
+      // between the time `dbList` is ran and `dbCreate` is.
+      if (error.message.match(/^Database `.*` already exists in/)) {
```

Comment:
As mentioned in the other thread, I think we can avoid the extra `r.dbList()` operation here since we catch the "already exists" error anyway.

There's also no need for the `Promise` constructor. We can just chain everything off of the promise returned from `r.dbCreate().run()`.